leap/multistep/__init__.py

```diff
@@ -385,35 +424,176 @@ def rk_bootstrap(self, cb):
 
-        # Stage loop (taken from EmbeddedButcherTableauMethodBuilder)
-        rhss = [var("rk_rhs_" + str(i)) for i in range(len(rk_tableau))]
-        for stage_num, (c, coeffs) in enumerate(rk_tableau):
-            if len(coeffs) == 0:
-                assert c == 0
-                cb(rhss[stage_num], rhs_var)
-            else:
-                stage = self.state + sum(self.dt * coeff * rhss[j]
-                                         for (j, coeff)
-                                         in enumerate(coeffs))
+        # Assign the value of the new state.
+        cb(self.state, est_vars[0])
+
+    def determine_bootstrap_length(self):
+
+        # In the explicit case, this is always
+        # equal to history length.
+        bootstrap_length = self.hist_length
+
+        return bootstrap_length
+# }}}
 
-                if self.state_filter is not None:
-                    stage = self.state_filter(stage)
 
-                cb(rhss[stage_num], self.eval_rhs(self.t + c * self.dt, stage))
+# {{{ am method
 
-        # Merge the values of the RHSs.
-        rk_comb = sum(coeff * rhss[j] for j, coeff in enumerate(rk_coeffs))
 
-        state_est = self.state + self.dt * rk_comb
+class AdamsMoultonMethodBuilder(AdamsMethodBuilder):
+    def generate_primary(self, cb):
+
+        from pytools import UniqueNameGenerator
+        name_gen = UniqueNameGenerator()
+        rhs_next_var = var("rhs_next_var")
+        rhs_var_to_unknown = {}
+        unkvar = cb.fresh_var("unk")
+        rhs_var_to_unknown[rhs_next_var] = unkvar
+
+        # In implicit mode, the time history must
+        # include the *next* point in time.
+        time_history_data, time_hist, \
+                dt_factor, t_end = self.set_up_time_history(cb, self.t + self.dt)
+
+        # Implicit setup - rhs_next_var is an unknown, needs implicit solve.
+        equations = []
+        unknowns = set()
+        knowns = set()
+
+        unknowns.add(rhs_next_var)
+
+        # Update history
+        history = self.history + [rhs_next_var]
+
+        # Set up the actual Adams-Moulton step.
+        am_sum = emit_adams_integration(
+                        cb, name_gen,
+                        self.function_family,
+                        time_hist, history,
+                        0, t_end)
+
+        state_est = self.state + dt_factor * am_sum
+
+        # Build the implicit solve expression.
+        from dagrt.expression import collapse_constants
+        from pymbolic.mapper.distributor import DistributeMapper as DistMap
+        solve_expression = collapse_constants(
+                rhs_next_var - self.eval_rhs(self.t + self.dt,
+                                                 DistMap()(state_est)),
+                list(unknowns) + [self.state],
+                cb.assign, cb.fresh_var)
+        equations.append(solve_expression)
+
+        # {{{ emit solve if possible
+
+        if unknowns and len(unknowns) == len(equations):
```

Comment:
What is the `else` to this if? Why is it OK to just go by here if we can't solve?

Reply:
The else - as far as I'm aware - would be the fully explicit case, but at bare minimum in that case the if condition isn't indicative of that.

Reply:
Hmm. I realize now, though, that I only have this in the Adams-Moulton class, so explicit isn't a concern. Let me think about contingencies/what should be done if we don't have a square system, or if it's even possible for that to happen with the current setup.

Reply:
Addressed by 3803398.

Reply:
Non-square systems aren't allowed IMO.

Reply:
When you say "not allowed" do you mean they should never happen? I added the elifs (and now an else) to this block to catch instances in which a system is non-square, but I'm not sure if you're saying that that's not sufficient, or if the check for a square system isn't needed at all because we should never *not* have one.